MaxText/train.py

```diff
@@ -267,9 +278,33 @@ def train_step(model, config, state, data, dropout_rng):
     rng2: A new rng key that can be used in future calls.
 
   """
-  train_loss_fn = functools.partial(loss_fn, model, config, data, dropout_rng, is_train=True)
-  grad_fn = jax.value_and_grad(train_loss_fn, has_aux=True)
-  (loss, aux), raw_grads = grad_fn(state.params)
+  if config.gradient_accumulation_steps > 1:
+    def accumulate_gradient(acc_grad_and_loss, data):
+      grad_func = jax.value_and_grad(loss_fn, argnums=4, has_aux=True)
+      (cur_batch_loss, aux), cur_batch_gradient = grad_func(model, config, data, dropout_rng, state.params, is_train=True)
+      accumulated_loss = cur_batch_loss + acc_grad_and_loss['loss']
+      accumulated_gradient = jax.tree_util.tree_map(lambda x, y: x + y, cur_batch_gradient, acc_grad_and_loss['grad'])
```

Comment:
Trying to understand more the logic here. Will we do the average after the sum? Or I missed it somewhere? Thanks!

Reply:
Instead of averaging gradients we scale the loss by loss = loss /accum_steps. This is equivalent to averaging the gradients

Reply:
Which step is this `loss /accum_steps`? others look good. Thanks!

Reply:
it happens in a different place - loss_fn: https://github.com/google/maxtext/pull/812/files#diff-45acd14896a500d4b0919c3e844732d7401e45bdf822f6d7c89a52d735bc9267R255

Reply:
Actually updated so it all happens in the same place, and weights by unpadded tokens across microbatches